aiida/tools/importexport/dbimport/readers.py

```diff
@@ -0,0 +1,328 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Archive reader classes."""
+from abc import ABC, abstractmethod
+import dataclasses
+import json
+import os
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
+import zipfile
+import tarfile
+
+from distutils.version import StrictVersion
+from wrapt import decorator
+
+from aiida.common.exceptions import InvalidOperation
+from aiida.common.folders import Folder, SandboxFolder
+from aiida.tools.importexport.common.config import EXPORT_VERSION, NODES_EXPORT_SUBFOLDER
+from aiida.tools.importexport.common.exceptions import (
+    CorruptArchive, ImportValidationError, IncompatibleArchiveVersionError
+)
+from aiida.tools.importexport.common.utils import export_shard_uuid
+
+
+def detect_file_type(in_path):
+    """For back-compatibility."""
+    if os.path.isdir(in_path):
+        return 'folder'
+    if tarfile.is_tarfile(in_path):
+        return 'tar'
+    if zipfile.is_zipfile(in_path):
+        return 'zip'
+    raise ImportValidationError(
+        'Unable to detect the input file format, it is neither a '
+        'tar file, nor a (possibly compressed) zip file.'
+    )
+
+
+@dataclasses.dataclass
+class ExportMetadata:
```

Comment:
do we need to replicate this or can it be reused between import and export?

Reply:
no indeed ideally these could/will be merged. I have just copied it here whilst I am working out the reader interface.